src/node_api.cc

```diff
@@ -705,6 +706,27 @@ bool FindWrapper(v8::Local<v8::Object> obj,
   return true;
 }
 
+std::map<v8::Isolate*, napi_env> napi_env_map;
+uv_once_t napi_env_once = UV_ONCE_INIT;
+uv_mutex_t napi_env_mutex;
+void init_mutex_once() {
+  CHECK_EQ(uv_mutex_init(&napi_env_mutex), 0);
+}
+napi_env get_env(v8::Isolate* isolate) {
+  napi_env result;
+
+  uv_once(&napi_env_once, init_mutex_once);
+
+  uv_mutex_lock(&napi_env_mutex);
+  result = napi_env_map[isolate];
+  if (result == nullptr) {
+    result = napi_env_map[isolate] = new napi_env__(isolate);
```

Comment:
We shall also need a `v8impl::release_env()` to dereference/delete. However, these only makes sense when we have a module unload strategy.